php-frontend/src/main/java/org/sonar/php/metrics/CognitiveComplexityVisitor.java

```diff
@@ -0,0 +1,318 @@
+    complexity.addComplexityWithNesting(tree.switchToken());
+
+    complexity.incNesting();
+    super.visitSwitchStatement(tree);
+    complexity.decNesting();
+  }
+
+  @Override
+  public void visitWhileStatement(WhileStatementTree tree) {
+    complexity.addComplexityWithNesting(tree.whileToken());
+
+    visit(tree.condition());
+    visitWithNesting(tree.statements());
+  }
+
+  @Override
+  public void visitDoWhileStatement(DoWhileStatementTree tree) {
+    complexity.addComplexityWithNesting(tree.doToken());
+
+    visitWithNesting(tree.statement());
+    visit(tree.condition());
+  }
+
+  @Override
+  public void visitForStatement(ForStatementTree tree) {
+    complexity.addComplexityWithNesting(tree.forToken());
+
+    tree.init().forEach(this::visit);
+    tree.condition().forEach(this::visit);
+    tree.update().forEach(this::visit);
+    visitWithNesting(tree.statements());
+  }
+
+  @Override
+  public void visitForEachStatement(ForEachStatementTree tree) {
+    complexity.addComplexityWithNesting(tree.foreachToken());
+
+    visit(tree.expression());
+    visit(tree.key());
+    visit(tree.value());
+    visitWithNesting(tree.statements());
+  }
+
+  @Override
+  public void visitCatchBlock(CatchBlockTree tree) {
+    complexity.addComplexityWithNesting(tree.catchToken());
+
+    complexity.incNesting();
+    super.visitCatchBlock(tree);
+    complexity.decNesting();
+  }
+
+  @Override
+  public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
+    complexity.incNesting();
+    super.visitFunctionDeclaration(tree);
+    complexity.decNesting();
+  }
+
+  @Override
+  public void visitFunctionExpression(FunctionExpressionTree tree) {
+    complexity.incNesting();
+    super.visitFunctionExpression(tree);
+    complexity.decNesting();
+  }
+
+  @Override
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    complexity.incNesting();
+    super.visitMethodDeclaration(tree);
+    complexity.decNesting();
+  }
+
+  @Override
+  public void visitConditionalExpression(ConditionalExpressionTree tree) {
+    complexity.addComplexityWithNesting(tree.queryToken());
+    visit(tree.condition());
+    visitWithNesting(tree.trueExpression());
+    visitWithNesting(tree.falseExpression());
+  }
+
+  @Override
+  public void visitBreakStatement(BreakStatementTree tree) {
+    if (tree.argument() != null) {
+      complexity.addComplexityWithoutNesting(tree.breakToken());
+    }
+    super.visitBreakStatement(tree);
+  }
+
+  @Override
+  public void visitContinueStatement(ContinueStatementTree tree) {
+    if (tree.argument() != null) {
+      complexity.addComplexityWithoutNesting(tree.continueToken());
+    }
+    super.visitContinueStatement(tree);
+  }
+
+  @Override
+  public void visitGotoStatement(GotoStatementTree tree) {
+    complexity.addComplexityWithoutNesting(tree.gotoToken());
+    super.visitGotoStatement(tree);
+  }
+
+  @Override
+  public void visitBinaryExpression(BinaryExpressionTree tree) {
```

Comment:
I suggest to first flatten the BinaryExpressionTree in a list of operators that reproduces the original syntax : `[AND,OR,AND]` and only after we have that list we calculate the complexity on it by scanning from left to right.
The algorithm to flatten could be something like :
`list = new_empty_list`
`i = 0`
`flatten(i, list, tree)`

`function flatten(i, list, tree)`
  `list.add(i,tree)`
  `flatten(i, list, tree.left)`
  `flatten(++i, list, tree.right)`
`end_function`